spark/dl/src/main/scala/com/intel/analytics/bigdl/parameters/AllReduceParameterManager.scala

```diff
@@ -0,0 +1,333 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.intel.analytics.bigdl.parameters
+
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.{Callable, Executors, Future, ThreadFactory}
+
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.{Engine, T, Table}
+import org.apache.commons.lang.exception.ExceptionUtils
+import org.apache.log4j.Logger
+import org.apache.spark.sparkExtension.{ParameterManagerMaster, SparkExtension}
+import org.apache.spark.SparkEnv
+import org.apache.spark.storage.{BlockId, BlockManagerWrapper, StorageLevel}
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable.HashMap
+import scala.reflect._
+
+object AllReduceParameterManager {
+  val logger = Logger.getLogger(getClass)
+
+  private val nextId = new AtomicInteger(0)
+
+  private val pm = new HashMap[String, AllReduceParameterManager]()
+
+  private var executorIdMap: HashMap[String, Int] = null
+
+  def setExecutorMap(map: HashMap[String, Int]): Unit = {
+    executorIdMap = map
+  }
+
+  def get(id: Int, executorId: String): AllReduceParameterManager = {
+    val eid = id + "exe" + executorIdMap(executorId)
+    if (pm.contains(eid)) pm(eid)
+    else null
+  }
+
+  def createParameterManager[T: ClassTag](executorId: Int, executorNum: Int, partitionNum: Int,
+    size: Int, port: Int = -1): AllReduceParameterManager = {
+    val id = nextId.getAndIncrement()
+    val conf = SparkEnv.get.conf
+    val master = ParameterManagerMaster.createEnv(conf, port)
+    val p = new AllReduceParameterManager(id, executorId, executorNum, partitionNum, size, master)
+    pm.put(id + "exe" + executorId, p)
+    p
+  }
+}
+
+/**
+ * Represent parameters stored on the block manager. In distributed optimization, we put parameters
+ * on block manager of spark. Each worker syncs parameters through the block manager. Block manager
+ * here serves as a parameter server.
+ *
+ * A Tensor is sliced into `executorNum` chunks and each chunk is assigned to a particular node
+ * (Spark executor). Likewise, gradients for each chunk are also assigned and stored on separate
+ * nodes. In this way, gradient aggregation and parameter updates can be performed independently for
+ * each chunk on separate nodes.
+ *
+ * @param id distinguish from other parameters
+ * @param executorId executorId of the node
+ * @param executorNum how many executors will use this parameter
+ * @param partitionNum partition numbers
+ * @param size size of the parameter (1D vector)
+ * @param master ParameterManagerMaster endpoint
+ */
+class AllReduceParameterManager(val id: Int, val executorId: Int, executorNum: Int,
+  partitionNum: Int, size: Int, val master: ParameterManagerMaster) {
+  import AllReduceParameterManager._
+
+  private val syncPoolSize: Int = System.getProperty(
```

Comment:
This part is copied from AllreduceParameter. Sync pool is a thread pool used to fetch parameters from other nodes and 4 is kind of a magic number that can achieve the best performance in our cluster. Since it's only used in pm and pool size is different with other pools in Engine. I left the code here.